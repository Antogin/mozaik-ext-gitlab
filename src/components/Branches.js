import React, { Component, PropTypes } from 'react'
import Branch                          from './Branch'
import {
    Widget,
    WidgetHeader,
    WidgetBody,
} from 'mozaik/ui'


export default class Branches extends Component {
    static propTypes = {
        project: PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.number,
        ]).isRequired,
        apiData: PropTypes.shape({
            project:  PropTypes.object,
            branches: PropTypes.array.isRequired,
        }).isRequired,
    }

    static defaultProps = {
        apiData: {
            project:  null,
            branches: [],
        },
    }

    static getApiRequest({ project }) {
        return {
            id:     `gitlab.projectBranches.${ project }`,
            params: { project }
        }
    }

    render() {
        const { apiData: { project, branches } } = this.props

        return (
            <Widget>
                <WidgetHeader
                    title="Project branches"
                    count={branches.length}
                    icon="code-fork"
                />
                <WidgetBody>
                    {branches.map(branch => (
                        <Branch key={branch.name} project={project} branch={branch} />
                    ))}
                </WidgetBody>
            </Widget>
        )
    }
}
