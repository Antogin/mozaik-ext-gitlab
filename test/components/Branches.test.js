import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import { WidgetHeader, WidgetLoader, defaultTheme } from '@mozaik/ui'
import Branches from './../../src/components/Branches'

const sampleProject = {
    name: 'ploucifier',
    web_url: 'https://gitlab.com/plouc/ploucifier',
}
const sampleBranches = [
    {
        name: 'fix-something',
        protected: false,
        created_at: '2015-10-22T11:38:08.000+02:00',
        commit: {
            id: '0610e6ffb1b8672eea4ba03591f1f0da13669bbc',
            message: 'Whatever',
            author_name: 'plouc',
            committed_date: '2015-10-22T11:38:08.000+02:00',
        },
    },
]

test('should return correct api request', () => {
    expect(
        Branches.getApiRequest({
            project: sampleProject.name,
        })
    ).toEqual({
        id: `gitlab.projectBranches.${sampleProject.name}`,
        params: { project: sampleProject.name },
    })
})

test('should display loader if no apiData available', () => {
    const wrapper = shallow(<Branches project={sampleProject.name} />)

    expect(wrapper.find(WidgetLoader).exists()).toBeTruthy()
})

test('header should display 0 count by default', () => {
    const wrapper = shallow(<Branches project={sampleProject.name} />)

    const header = wrapper.find(WidgetHeader)
    expect(header.prop('count')).toBe(0)
})

test('header should display pull request count when api sent data', () => {
    const wrapper = shallow(
        <Branches
            project={sampleProject.name}
            apiData={{ project: sampleProject, branches: sampleBranches }}
        />
    )

    const header = wrapper.find(WidgetHeader)
    expect(header.exists()).toBeTruthy()
    expect(header.prop('count')).toBe(sampleBranches.length)
})

test(`header title should default to '<project_name> Branches'`, () => {
    const wrapper = shallow(
        <Branches
            project={sampleProject.name}
            apiData={{ project: sampleProject, branches: sampleBranches }}
        />
    )

    const header = wrapper.find(WidgetHeader)
    expect(header.prop('title')).toBe('Branches')
})

test(`header title should be overridden when passing 'title' prop`, () => {
    const customTitle = 'Custom Title'
    const wrapper = shallow(
        <Branches
            project={sampleProject.name}
            apiData={{ project: sampleProject, branches: sampleBranches }}
            title={customTitle}
        />
    )

    const header = wrapper.find(WidgetHeader)
    expect(header.prop('title')).toBe(customTitle)
})

test('should render as expected', () => {
    const tree = renderer.create(
        <ThemeProvider theme={defaultTheme}>
            <Branches
                project={sampleProject.name}
                apiData={{ project: sampleProject, branches: sampleBranches }}
            />
        </ThemeProvider>
    )

    expect(tree).toMatchSnapshot()
})
