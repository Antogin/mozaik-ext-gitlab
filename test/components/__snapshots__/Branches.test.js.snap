// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render as expected 1`] = `
<div
  className="sc-bdVaJa eydJRb"
  style={Object {}}
>
  <div
    className="sc-bwzfXH fMILzD"
  >
    <header
      className="sc-htpNat llvRri"
      style={Object {}}
    >
      <span>
        <div
          className="sc-ifAKCX bDEKRH"
        >
          <a
            href="https://gitlab.com/plouc/ploucifier"
            target="_blank"
          >
            ploucifier
          </a>
        </div>
        Branches
        <div
          className="sc-bxivhb eDwDKI"
        >
          1
        </div>
      </span>
      <div
        className="sc-EHOje dlEdwN"
      >
        <svg
          fill="currentColor"
          height="1em"
          preserveAspectRatio="xMidYMid meet"
          style={
            Object {
              "color": "#000",
              "verticalAlign": "middle",
            }
          }
          viewBox="0 0 40 40"
          width="1em"
        >
          <g>
            <path
              d="m14.9 32.9q0-0.9-0.6-1.6t-1.5-0.6-1.5 0.6-0.7 1.6 0.7 1.5 1.5 0.6 1.5-0.6 0.6-1.5z m0-25.8q0-0.8-0.6-1.5t-1.5-0.6-1.5 0.6-0.7 1.5 0.7 1.6 1.5 0.6 1.5-0.6 0.6-1.6z m14.3 2.9q0-0.9-0.6-1.5t-1.5-0.6-1.5 0.6-0.7 1.5 0.7 1.5 1.5 0.6 1.5-0.6 0.6-1.5z m2.2 0q0 1.2-0.6 2.2t-1.6 1.5q0 6.4-5 9.2-1.6 0.9-4.6 1.9-2.8 0.8-3.7 1.5t-1 2.3v0.5q1 0.6 1.6 1.6t0.6 2.2q0 1.7-1.3 3t-3 1.2-3.1-1.2-1.2-3q0-1.2 0.6-2.2t1.5-1.5v-18.4q-0.9-0.5-1.5-1.5t-0.6-2.2q0-1.7 1.3-3t3-1.2 3 1.2 1.3 3q0 1.2-0.6 2.2t-1.6 1.5v11.1q1.2-0.5 3.5-1.2 1.2-0.4 1.9-0.7t1.6-0.7 1.3-0.9 0.9-1.1 0.6-1.6 0.2-2q-1-0.6-1.5-1.5t-0.6-2.2q0-1.8 1.2-3t3.1-1.3 3 1.3 1.3 3z"
            />
          </g>
        </svg>
      </div>
    </header>
    <div
      className="sc-bZQynM cUinle"
      style={Object {}}
    >
      <div>
        <div
          className="sc-cSHVUG gWizrO"
          onClick={undefined}
          style={undefined}
        >
          <div
            style={
              Object {
                "flexGrow": 1,
              }
            }
          >
            <span>
              <a
                href="https://gitlab.com/plouc/ploucifier/tree/fix-something"
                target="_blank"
              >
                fix-something
              </a>
               
              <a
                href="https://gitlab.com/plouc/ploucifier/commit/0610e6ffb1b8672eea4ba03591f1f0da13669bbc"
                style={
                  Object {
                    "textDecoration": "underline",
                  }
                }
                target="_blank"
              >
                #
                0610e6f
              </a>
            </span>
            <div
              className="sc-kgoBCf kFsOlB"
            >
              <div>
                <div>
                  Whatever
                </div>
                <time
                  style={
                    Object {
                      "alignItems": "center",
                      "display": "flex",
                    }
                  }
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    preserveAspectRatio="xMidYMid meet"
                    style={
                      Object {
                        "color": undefined,
                        "verticalAlign": "middle",
                      }
                    }
                    viewBox="0 0 40 40"
                    width="1em"
                  >
                    <g>
                      <path
                        d="m23 12.1v10q0 0.4-0.2 0.6t-0.5 0.2h-7.2q-0.3 0-0.5-0.2t-0.2-0.6v-1.4q0-0.3 0.2-0.5t0.5-0.2h5v-7.9q0-0.3 0.2-0.5t0.6-0.2h1.4q0.3 0 0.5 0.2t0.2 0.5z m9.3 7.9q0-3.3-1.6-6.1t-4.5-4.4-6.1-1.6-6.1 1.6-4.4 4.4-1.6 6.1 1.6 6.1 4.4 4.4 6.1 1.6 6.1-1.6 4.5-4.4 1.6-6.1z m5 0q0 4.7-2.3 8.6t-6.3 6.2-8.6 2.3-8.6-2.3-6.2-6.2-2.3-8.6 2.3-8.6 6.2-6.2 8.6-2.3 8.6 2.3 6.3 6.2 2.3 8.6z"
                      />
                    </g>
                  </svg>
                   
                  2 years ago
                </time>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
